packages/flutter/lib/src/painting/fractional_offset.dart

```diff
@@ -193,5 +212,86 @@ class FractionalOffset {
   }
 
   @override
+  FractionalOffset resolve(TextDirection direction) => this;
+
+  @override
+  bool operator ==(dynamic other) {
+    if (other.runtimeType != runtimeType)
+      return false;
+    final FractionalOffset typedOther = other;
+    return dx == typedOther.dx &&
+           dy == typedOther.dy;
+  }
+
+  @override
+  int get hashCode => hashValues(dx, dy);
+
+  @override
   String toString() => '$runtimeType($dx, $dy)';
 }
+
+/// An offset that's expressed as a fraction of a [Size], but whose horizontal
+/// component is dependent on the writing direction.
+///
+/// This can be used to indicate an offset from the left in [TextDirection.ltr]
+/// text and an offset from the right in [TextDirection.rtl] text without having
+/// to be aware of the current text direction.
+class FractionalOffsetDirectional extends FractionalOffsetGeometry {
+  /// Creates a directional fractional offset.
+  ///
+  /// The [start] and [dy] arguments must not be null.
+  const FractionalOffsetDirectional(this.start, this.dy)
+    : assert(start != null),
+      assert(dy != null);
+
+  /// The distance fraction in the horizontal direction.
+  ///
+  /// A value of 0.0 corresponds to the edge on the "start" side, which is the
+  /// left side in [TextDirection.ltr] contexts and the right side in
+  /// [TextDirection.rtl] contexts. A value of 1.0 corresponds to the opposite
+  /// edge, the "end" side. Values are not limited to that range; negative
+  /// values represent positions beyond the start edge, and values greater than
+  /// 1.0 represent positions beyond the end edge.
+  ///
+  /// This value is normalized into a [FractionalOffset.dx] value by the
+  /// [resolve] method.
+  final double start;
+
+  /// The distance fraction in the vertical direction.
+  ///
+  /// A value of 0.0 corresponds to the topmost edge. A value of 1.0 corresponds
+  /// to the bottommost edge. Values are not limited to that range; negative
+  /// values represent positions above the top, and values greater than 1.0
+  /// represent positions below the bottom.
+  ///
+  /// This value is passed through to [FractionalOffset.dy] unmodified by the
+  /// [resolve] method.
+  final double dy;
+
```

Comment:
In the spirit of "just in time" problem solving, I'm testing the hypothesis that one only actually does maths with these once they're in the absolute coordinate space (e.g. just before painting, etc).

I suspect some static const values will make sense, but I'm similarly waiting to see what the needs are before adding them.